File: app/api/v01/vehicle_usage_sets.rb
# Copyright © Mapotempo, 2015
#
# This file is part of Mapotempo.
#
# Mapotempo is free software. You can redistribute it and/or
# modify since you respect the terms of the GNU Affero General
# Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
#
# Mapotempo is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the Licenses for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Mapotempo. If not, see:
# <http://www.gnu.org/licenses/agpl.html>
#
require 'coerce'

class V01::VehicleUsageSets < Grape::API
  helpers SharedParams
  helpers do
    # Never trust parameters from the scary internet, only allow the white list through.
    def vehicle_usage_set_params
      p = ActionController::Parameters.new(params)
      p = p[:vehicle_usage_set] if p.key?(:vehicle_usage_set)
      p.permit(:name, :open, :close, :store_start_id, :store_stop_id, :service_time_start, :service_time_end, :rest_start, :rest_stop, :rest_duration, :store_rest_id)
    end
  end

  resource :vehicle_usage_sets do
    desc 'Fetch customer\'s vehicle_usage_sets. At least one vehicle_usage_set exists per customer.',
      nickname: 'getVehicleUsageSets',
      is_array: true,
      success: V01::Entities::VehicleUsageSet
    params do
      optional :ids, type: Array[Integer], desc: 'Select returned vehicle_usage_sets by id.', coerce_with: CoerceArrayInteger
    end
    get do
      vehicle_usage_sets = if params.key?(:ids)
        current_customer.vehicle_usage_sets.select{ |vehicle_usage_set| params[:ids].include?(vehicle_usage_set.id) }
      else
        current_customer.vehicle_usage_sets.load
      end
      present vehicle_usage_sets, with: V01::Entities::VehicleUsageSet
    end

    desc 'Fetch vehicle_usage_set.',
      nickname: 'getVehicleUsageSet',
      success: V01::Entities::VehicleUsageSet
    params do
      requires :id, type: Integer
    end
    get ':id' do
      present current_customer.vehicle_usage_sets.find(params[:id]), with: V01::Entities::VehicleUsageSet
    end

    desc 'Create vehicle_usage_set.',
      detail: "(Note a vehicle_usage_set is already automatically created with a customer.) Only available if \"multi usage set\" option is active for current customer.
        For instance, if customer needs to use its vehicle 2 times per day (morning and evening), he needs 2 VehicleUsageSet called \"Morning\" and \"Evening\". A VehicleUsage per Vehicle is automatically created. The new VehicleUsageSet allows to define new default values for VehicleUsage.",
      nickname: 'createVehicleUsageSet',
      success: V01::Entities::VehicleUsageSet
    params do
      use :params_from_entity, entity: V01::Entities::VehicleUsageSet.documentation.except(
          :id,
          :open,
          :close,
          :service_time_start,
          :service_time_end,
          :rest_start,
          :rest_stop,
          :rest_duration
      ).deep_merge(
        name: { required: true },
        store_start_id: { required: true },
        store_stop_id: { required: true }
      )

      requires :open, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      requires :close, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :service_time_start, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :service_time_end, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :rest_start, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :rest_stop, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :rest_duration, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
    end
    post do
      vehicle_usage_set = current_customer.vehicle_usage_sets.build(vehicle_usage_set_params)
      vehicle_usage_set.save!
      present vehicle_usage_set, with: V01::Entities::VehicleUsageSet
    end

    desc 'Update vehicle_usage_set.',
      nickname: 'updateVehicleUsageSet',
      success: V01::Entities::VehicleUsageSet
    params do
      requires :id, type: Integer
      use :params_from_entity, entity: V01::Entities::VehicleUsageSet.documentation.except(
          :id,
          :open,
          :close,
          :service_time_start,
          :service_time_end,
          :rest_start,
          :rest_stop,
          :rest_duration)

      optional :open, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :close, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :service_time_start, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :service_time_end, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :rest_start, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :rest_stop, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
      optional :rest_duration, type: Integer, documentation: { type: 'string', desc: 'Schedule time (HH:MM)' }, coerce_with: ->(value) { ScheduleType.new.type_cast(value) }
    end
    put ':id' do
      vehicle_usage_set = current_customer.vehicle_usage_sets.find(params[:id])
      vehicle_usage_set.update! vehicle_usage_set_params
      present vehicle_usage_set, with: V01::Entities::VehicleUsageSet
    end

    desc 'Delete vehicle_usage_set.',
      nickname: 'deleteVehicleUsageSet'
    params do
      requires :id, type: Integer
    end
    delete ':id' do
      current_customer.vehicle_usage_sets.find(params[:id]).destroy
      status 204
    end

    desc 'Delete multiple vehicle_usage_sets.',
      nickname: 'deleteVehicleUsageSets'
    params do
      requires :ids, type: Array[Integer], coerce_with: CoerceArrayInteger
    end
    delete do
      VehicleUsageSet.transaction do
        current_customer.vehicle_usage_sets.select{ |vehicle_usage_set| params[:ids].include?(vehicle_usage_set.id) }.each(&:destroy)
      end
      status 204
    end

    desc 'Import vehicle usage set by upload a CSV file or by JSON.',
         nickname: 'importVehicleUsageSets',
         is_array: true,
         success: V01::Entities::VehicleUsageSet
    params do
      use :params_from_entity, entity: V01::Entities::VehicleUsageSetsImport.documentation
    end
    put do
      import = if params[:vehicle_usage_sets]
                 ImportJson.new(importer: ImporterVehicleUsageSets.new(current_customer), replace_vehicles: params[:replace_vehicles], json: params[:vehicle_usage_sets])
               else
                 ImportCsv.new(importer: ImporterVehicleUsageSets.new(current_customer), replace_vehicles: params[:replace_vehicles], file: params[:file])
               end

      if import && import.valid? && (vehicles_usages_with_conf = import.import)
        present vehicles_usages_with_conf, with: V01::Entities::Vehicle
      else
        error!({error: import.errors.full_messages}, 422)
      end
    end
  end
end

File: app/assets/javascripts/vehicle_usage_sets.js
// Copyright © Mapotempo, 2013-2017
//
// This file is part of Mapotempo.
//
// Mapotempo is free software. You can redistribute it and/or
// modify since you respect the terms of the GNU Affero General
// Public License as published by the Free Software Foundation,
// either version 3 of the License, or (at your option) any later version.
//
// Mapotempo is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE.  See the Licenses for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Mapotempo. If not, see:
// <http://www.gnu.org/licenses/agpl.html>
//
'use strict';

var vehicle_usage_sets_index = function(params) {
  'use strict';

  // override accordion collapse bootstrap code
  $('a.accordion-toggle').click(function() {
    var id = $(this).attr('href');
    window.location.hash = id;
    $('.accordion-body.collapse.in').each(function(index) {
      var $this = $(this);
      if (id !== '#' + $this.attr('id')) {
        $this.collapse('hide');
      }
    });
  });

  if (window.location.hash) {
    $('.accordion-body.collapse.in').each(function(index) {
      var $this = $(this);
      if (window.location.hash !== '#' + $this.attr('id')) {
        $this.removeClass('in');
      }
    });
    $(".accordion-toggle[href!='" + window.location.hash + "']").addClass('collapsed');
    $(window.location.hash).addClass('in');
    $(".accordion-toggle[href='" + window.location.hash + "']").removeClass('collapsed');
  }
};

var vehicle_usage_sets_edit = function(params) {
  $('#vehicle_usage_set_open, #vehicle_usage_set_close, #vehicle_usage_set_rest_start, #vehicle_usage_set_rest_stop, #vehicle_usage_set_rest_duration, #vehicle_usage_set_service_time_start, #vehicle_usage_set_service_time_end').timeEntry({
    show24Hours: true,
    spinnerImage: '',
    defaultTime: '00:00'
  });
};

var vehicle_usage_sets_import = function(params) {
  'use strict';

  var dialog_upload = bootstrap_dialog({
    title: I18n.t('vehicle_usage_sets.import.dialog.import.title'),
    icon: 'fa-upload',
    message: SMT['modals/default_with_progress']({
      msg: I18n.t('vehicle_usage_sets.import.dialog.import.in_progress')
    })
  });

  $(":file").filestyle({
    buttonName: "btn-primary",
    iconName: "fa fa-folder-open",
    buttonText: I18n.t('web.choose_file')
  });

  $('form#new_import_csv').submit(function () {
    var confirmChecks = [];
    $('#import_csv_replace_vehicles', $(this)).is(':checked') && confirmChecks.push('replace_vehicles');
    if (confirmChecks.length > 0 && !confirm(confirmChecks.map(function (c) {
        var vehicle_usage_set_import_translation = 'vehicle_usage_sets.import.dialog.' + c + '_confirm';
        return I18n.t(vehicle_usage_set_import_translation);
      }).join(" \n"))) {
      return false;
    }

    dialog_upload.modal(modal_options());
  });
};

Paloma.controller('VehicleUsageSets', {
  index: function() {
    vehicle_usage_sets_index(this.params);
  },
  new: function() {
    vehicle_usage_sets_edit(this.params);
  },
  create: function() {
    vehicle_usage_sets_edit(this.params);
  },
  edit: function() {
    vehicle_usage_sets_edit(this.params);
  },
  update: function() {
    vehicle_usage_sets_edit(this.params);
  },
  import: function() {
    vehicle_usage_sets_import(this.params);
  },
  upload_csv: function() {
    vehicle_usage_sets_import(this.params);
  }
});
